commands/dice.js

```diff
@@ -2,37 +2,133 @@ var Roll = require('roll');
 
 exports.run = function(config, msg) {
   var roll = new Roll();
-  var valid = roll.validate(msg.content);
   var input = msg.content;
-  if(valid) {
-    var dicebag = input.split("+");
-    for(var i = 0; i < dicebag.length; i++) {
-      console.log(dicebag[i]);
-      var isDie = isNaN(dicebag[i]);
-      if(isDie) {
-        //check for modifiers to be added to the die
-        for(var j = i+1; j < dicebag.length; j++) {
-          console.log(dicebag[j]);
-          if(!isNaN(dicebag[j])) {
-            dicebag[i] = dicebag[i]+"+"+dicebag[j];
-          } else {
-            break;
+  var dicebag = input.split("+");
+  var output;
+
+  dicebag = assembleDice(dicebag);
+  for(var i = 0; i < dicebag.length; i++) {
+    var advantage = "none";
+
+    //check for advantage
+    if(dicebag[i].includes("A")) {
+      advantage = "advantage";
+      dicebag[i] = dicebag[i].replace("A", "");
+    } else if (dicebag[i].includes("D")) {
+      advantage = "disadvantage";
+      dicebag[i] = dicebag[i].replace("D", "");
+    }
+
+    var valid = roll.validate(dicebag[i]);
+    if(valid) {
+      //collect some data before rolling
+      var count,
+          type;
+      var modifiers = [];
+
+      if(dicebag[i].match(/^\d/)) {
+        count = dicebag[i].split("d")[0];
+      } else {
+        count = 0;
+      }
+
+      var temp = dicebag[i].split("d")[1]
+      for(var j=0; j<temp.length; j++){
+        if(isNaN(temp.charAt(j))) {
+          if(type === undefined) {
+            type = temp.substring(0, j);
+          }
+          for(var k=j+1; k <temp.length; k++) {
+            if(isNaN(temp.charAt(k))) {
+              modifiers.push({value: temp.substring(j+1, k), sign: temp.charAt(j)});
+              break;
+            }
+
+            if(k === (temp.length - 1)){
+              modifiers.push({value: temp.substring(j+1), sign: temp.charAt(j)});
+            }
+          }
+        }
+      }
+
+      //determine what kind of roll we are making
+      if(advantage != "none") {
+        var temp = (count > 0)? dicebag[i].replace(dicebag[i].charAt(0), "2"): "2"+dicebag[i];
```

Comment:
This can be a big problem by blindly changing the first char in the string to do. What happens if its 33d6? we now roll 23d6 which is just odd and not right.